File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # approach: stack
        
        if len(tokens) == 1:
            return int(tokens[0])
        
        ops = {'+', '-', '*', '/'}
        stack = []
        
        for t in tokens:
            if t not in ops:
                stack.append(int(t))
            else:
                right, left = stack.pop(), stack.pop()
                
                if t == '+':
                    stack.append(left + right)
                elif t == '-':
                    stack.append(left - right)
                elif t == '*':
                    stack.append(left * right)
                else:
                    stack.append(int(left / right))
                    
        return stack[0]

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        Set<String> ops = new HashSet<>();
        
        ops.add("+");
        ops.add("-");
        ops.add("*");
        ops.add("/");

        for (String str : tokens) {
            if (ops.contains(str)) {
                // handle operator
                int right = stack.pop();
                int left = stack.pop();
                
                if (str.equals("+")) {
                    stack.push(left + right);
                } else if (str.equals("-")) {
                    stack.push(left - right);
                } else if (str.equals("*")) {
                    stack.push(left * right);
                } else {
                    stack.push(left / right);
                }
            } else {
                // handle num
                int num = Integer.valueOf(str);
                stack.push(num);
            }
        }
        
        // last element is result
        return stack.pop();
    }
}